node-repository/src/main/java/com/yahoo/vespa/hosted/provision/provisioning/GroupPreparer.java

```diff
@@ -59,28 +61,26 @@ public GroupPreparer(NodeRepository nodeRepository,
                               List<Node> surplusActiveNodes, MutableInteger highestIndex, int wantedGroups) {
         boolean dynamicProvisioningEnabled = nodeRepository.canProvisionHosts() && nodeRepository.zone().getCloud().dynamicProvisioning();
         boolean allocateFully = dynamicProvisioningEnabled && preprovisionCapacityFlag.value().isEmpty();
-        try (Mutex lock = nodeRepository.lock(application)) {
 
-            // Lock ready pool to ensure that the same nodes are not simultaneously allocated by others
-            try (Mutex allocationLock = nodeRepository.lockUnallocated()) {
+        // Try preparing in memory without lock. Most of the time there should be no changes and we can return nodes
+        // previously allocated.
+        {
+            MutableInteger probePrepareHighestIndex = new MutableInteger(highestIndex.get());
+            NodeAllocation probeAllocation = prepareAllocation(application, cluster, requestedNodes, surplusActiveNodes,
+                    probePrepareHighestIndex, wantedGroups, allocateFully, PROBE_LOCK);
+            if (probeAllocation.fulfilledAndNoChanges()) {
+                List<Node> acceptedNodes = probeAllocation.finalNodes();
+                surplusActiveNodes.removeAll(acceptedNodes);
```

Comment:
Why are nodes removed from surplusActiveNodes - I don't see any nodes are removed in the current/previous code?

Reply:
The same thing is done in the existing code, at the end of the method.

Reply:
Now I see it - never mind

Reply:
Removed here:
https://github.com/vespa-engine/vespa/blob/90c8b5c969f16ba7ca0bc64f718a3ad8bdf527ff/node-repository/src/main/java/com/yahoo/vespa/hosted/provision/provisioning/GroupPreparer.java#L119

Reply:
Well, under the condition of dynamicProvisionEnabled